mountpoint-s3-client/src/s3_crt_client/put_object.rs

```diff
@@ -106,6 +126,47 @@ pub struct S3PutObjectRequest {
     review_callback: ReviewCallbackBox,
     start_time: Instant,
     total_bytes: u64,
+    // Headers of the CompleteMultipartUpload response, available after the request was finished
+    response_headers: Arc<Mutex<Option<Headers>>>,
+    // Server-side encryption type which is expected to be found in response_headers
+    pub server_side_encryption: Option<String>,
+    /// Server-side encryption KMS key ID which is expected to be found in response_headers
+    pub ssekms_key_id: Option<String>,
+}
+
+/// If non empty `server_side_encryption` or `ssekms_key_id` were used, this function checks headers
+/// of the CompleteMultipartUpload response to contain the expected values
+fn check_response_headers(
+    response_headers: Arc<Mutex<Option<Headers>>>,
+    expected_sse: &Option<String>,
+    expected_key_id: &Option<String>,
+) {
+    let locked_value = response_headers.lock().expect("must be able to acquire headers lock");
+    let actual_headers = locked_value
+        .as_ref()
+        .expect("PUT response headers must be available at this point");
+    if let Some(sse_type) = expected_sse {
+        assert!(
+            actual_headers.get(SSE_TYPE_HEADER_NAME).is_ok_and(|header| {
+                header
+                    .value()
+                    .to_str()
+                    .is_some_and(|actual_value| actual_value == sse_type)
+            }),
+            "SSE type provided in CompleteMultipartUpload response does not match the requested value"
+        );
+    }
```

Comment:
nit: could you refactor these into `assert_eq!` calls and get a nicer output? e.g. **UNTESTED!**

```suggestion
    if let Some(sse_type) = expected_sse {
        let actual_header = actual_headers.get(SSE_TYPE_HEADER_NAME).ok();
        let actual_value = actual_header.as_ref().and_then(|header| header.value().to_str());
        assert_eq!(
            actual_value,
            Some(sse_type),
            "SSE type provided in CompleteMultipartUpload response does not match the requested value"
        );
    }
```

Reply:
actually this worked as is, looks better, applied the same for the `SSE_KEY_ID_HEADER_NAME`